### PR TITLE
Feature: kafka trigger

### PR DESCRIPTION
**Pull request by alexrodfe**
## Description

- Added the new kafka trigger
- Added documentation on usage
- Added config files needed for local deploy

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other changes (ci configuration, documentation or any other kind of changes)

## Checklist:

- [ ] I have created tests for my code changes, and the tests are passing.
- [ ] I have executed the pre-commit hooks locally.
- [ ] My change requires a change to the documentation (create a new issue if the documentation has not been updated).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: kafka-trigger/README.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+The configuration should be defined inside the `centralized configuration scope`:
+
+| Key         | Optional  | Type | Value                                       |
+|-------------|-----------|------|---------------------------------------------|
+| brokers     | no        | str  | Brokers' addresses (comma separated value)  |
+| groupid     | no        | str  | The groupID the listener will take          |
+| topic       | no        | str  | The topic's name                            |
+| tls_enabled | yes        | bool  | Enable TLS connection                     |
+| insecure_skip_verify | yes        | bool  | Skip SSL certificate validation  |
```
Comment (dfinteligenz) [inline kafka-trigger/README.md:28]:
```suggestion
| insecure_skip_verify | yes        | bool  | Skip SSL certificate validation, defaults to false  |
```

File: kafka-trigger/README.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+```
+
+### Output
+
+The trigger upon receiving a message, will send a key-value protobuf message with the following format:
+
+| Key       | Type    | Value                    |
+|-----------|---------|--------------------------|
+| topic     | str     | The topic's name         |
+| message   | []byte  | The message's payload    |
```
Comment (dfinteligenz) [inline kafka-trigger/README.md:49]:
```suggestion
| payload   | []byte  | The event's payload    |
```

File: kafka-trigger/sonar-project.properties
Comment (dfinteligenz) [inline kafka-trigger/sonar-project.properties:1]: You need to modify the CI to execute the tests and upload the results to Sonar.

If you need me to create a new project for this trigger, please just ask me.

**Review by dfinteligenz — CHANGES_REQUESTED** [review]

File: kafka-trigger/README.md (hunk 1 above)
Comment (dfinteligenz) [reply kafka-trigger/README.md:28]: Id' also rename the `insecure_skip_verify` to `skip_tls_verify`, IMO it's easier to understand.

File: kafka-trigger/kafkacluster.go
Comment (dfinteligenz) [inline kafka-trigger/kafkacluster.go:1]: Do we still need this file @alexrodfe , given that we are using the testcontainer predefined image

**Review by dfinteligenz — CHANGES_REQUESTED** [review]
Check comments above

File: kafka-trigger/kafkacluster.go
Comment (alexrodfe) [reply kafka-trigger/kafkacluster.go:1]: No we don't. I'll delete it

File: kafka-trigger/main.go [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+	)
+}
+
+func kafkaRunner(tr *trigger.Runner, kaiSDK sdk.KaiSDK) {
+	var err error
+
+	var dialer *kafka.Dialer
+	if config.TLSEnabled {
+		dialer = &kafka.Dialer{
+			Timeout:   10 * time.Second,
```
Comment (JYisus) [inline kafka-trigger/main.go:101]: Maybe this could be defined in config or as a constant

File: kafka-trigger/main.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+			)
+
+			err = sendMessage(kaiSDK, config.Topic, m.Value)
+			if err != nil {
+				kaiSDK.Logger.Error(err, "error sending message")
+				break
+			}
+		}
+
+		err := r.Close() // reader must be gracefully closed
```
Comment (JYisus) [inline kafka-trigger/main.go:140]: To ensure this is always called, call it with a defer in the first line of the goroutine

File: kafka-trigger/main.go [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+		}
+		os.Exit(1)
+	}()
+
+	// Handle sigterm and await termChan signal
+	termChan := make(chan os.Signal, 1)
+	signal.Notify(termChan, syscall.SIGINT, syscall.SIGTERM)
+	<-termChan
+
+	err = r.Close() // reader must be gracefully closed
```
Comment (JYisus) [inline kafka-trigger/main.go:152]: This is already been called in the goroutine, isn't it?

File: kafka-trigger/main.go [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+		err := r.Close() // reader must be gracefully closed
+		if err != nil {
+			kaiSDK.Logger.Error(err, "error closing reader")
+		}
+		os.Exit(1)
+	}()
+
+	// Handle sigterm and await termChan signal
+	termChan := make(chan os.Signal, 1)
```
Comment (JYisus) [inline kafka-trigger/main.go:148]: IMO, this logic to handle external signals to exit the program should be in the main function

**Review by JYisus — CHANGES_REQUESTED** [review]
Great job! Just a few things.

File: kafka-trigger/main.go (hunk 6 above)
Comment (alexrodfe) [reply kafka-trigger/main.go:143]: Main function launches runner, runner wakes up two go functions (runner and subscriber) then waits for them, subscriber stands waiting for sigterm. Runner should also be waiting for sigterm, if it wasnt, runner would end execution, then closing the go rutine inside

File: kafka-trigger/main.go (hunk 5 above)
Comment (alexrodfe) [reply kafka-trigger/main.go]: Yes but it was here to complete all posible scenarios, I'll add a defer as referred previously, so this line will no longer be needed

File: kafka-trigger/main.go (hunk 4 above)
Comment (alexrodfe) [reply kafka-trigger/main.go]: I wasn't adding a defer so I could check on the error but I agree, defer is much cleaner

File: kafka-trigger/main.go (hunk 3 above)
Comment (alexrodfe) [reply kafka-trigger/main.go]: I'll have it as a constant

**Review by dfinteligenz — APPROVED** [review]

**Review by JYisus — APPROVED** [review]

**Pull request merged**